wip-asdpc-rng.md

```diff
@@ -0,0 +1,182 @@
+  Status: Draft
+  Type: Standards Track
+  Created: 2021-09-06
+  License: BSD-2-Clause
+</pre>
+
+
+## Abstract
+
+A new type of data source is proposed such that leverages the crowd-attestation mechanism of the Witnet protocol
+to enable Random Number Generation (RNG) functionality inside data requests.
+
+## Motivation and rationale
+
+Just in the same way as price feeds are the fundamental oracle-based building block for many DeFi protocols RNGs are
+quickly becoming a keystone building block for non-fungible tokens (NFTs).  
+
+Many use cases are arising in which the smart contracts generating the NFTs are in need of reliable and tamper-proof
+random inputs for their generative algorithms ensuring the "uniqueness" of each piece. That is, they need to randomly
+assign a distinguishable set of scarce attributes or traits to each newly minted NFT unit. 
+
+The case for oracle-powered RNGs expands beyond NFTs, as many other constructs also benefit from random inputs, e.g.
+in the area of gaming, lotteries, gambling, random committee selection, etc.
+
+Due to the robustness and flexibility of the crowd-attestation mechanism of the Witnet protocol, there is a huge
+opportunity to add RNG capabilities to it with very little and self-contained changes.
+
+In particular, the mechanism hereby introduced consists on a new type of data source in which no external HTTP(S)
+servers are queried, and nodes are instead required to secretly commit secret sequences of bytes. The rest of the
+protocol follows in the traditional way, with the only addition of a new reducer function that concatenates the
+revealed values in a deterministic order and hashes the sequence together so as to produce a single, uniform and
+pseudo-random output that will be used as the result of the request.
+
+Additionally, a new deterministic ordering for reveals inside tally transactions is introduced so as to rule out any
+malleability attempt from tally producers. 
+
+Due to the overlap of Witnet's inherent commitment scheme with this last aggregation step, there is no need for the
+committed bytes to be verifiably random. Under the assumptions of stock hashing functions, even if the committers
+collude around crafting their commitments in such a way that the output is favorable to them, as long as 1 single
+committer remains independent and commits a value that is secret to the rest, the final result will hold the desired
+randomness properties.
+
+## Specification
+
+### Introduction of data request type `1`
+
+**(1)** A new type of data source is introduced with name `Rng` and identifier `1`.
+
+**(2)** Type `1` sources have the same `script` attribute as type `0` sources, but no `url` attribute.
+
+**(3)** The script field in type `1` sources use the same RADON operators and encoding as type `0`.
+
+### Retrieval of type `1` sources
+
+**(4)** When retrieving a type `1` source, a node MUST generate on the spot a random and secret sequence of 32
```

Comment:
Technically, the protocol admit a secuence of bytes different of 32, but it will be handle in Tally creation to adapt to 32 bytes (zeropadding or cutting)

Reply:
Still they MUST do so. What they are free to do is to ignore the MUST :rofl:

Reply:
I would be ok with changing it for SHOULD to acknowledge that there's no way to enforce it.